ts/index.ts

```diff
@@ -467,6 +497,65 @@ class Gossipsub extends BasicPubsub {
         this.log('PRUNE: Remove mesh link to %s in %s', peer.id.toB58String(), topicID)
         peers.delete(peer)
         peer.topics.delete(topicID)
+        // is there a backoff specified by the peer? if so obey it
+        if (typeof backoff === 'number' && backoff > 0) {
+          this._doAddBackoff(id, topicID, backoff * 1000)
+        } else {
+          this._addBackoff(id, topicID)
+        }
+      }
+    })
+  }
+
+  /**
+   * Add standard backoff log for a peer in a topic
+   * @param {string} id
+   * @param {string} topic
+   * @returns {void}
+   */
+  _addBackoff (id: string, topic: string): void {
+    this._doAddBackoff(id, topic, constants.GossipsubPruneBackoff)
+  }
+
+  /**
+   * Add backoff expiry interval for a peer in a topic
+   * @param {string} id
+   * @param {string} topic
+   * @param {number} interval backoff duration in milliseconds
+   * @returns {void}
+   */
+  _doAddBackoff (id: string, topic: string, interval: number): void {
+    let backoff = this.backoff.get(topic)
+    if (!backoff) {
+      backoff = new Map()
+      this.backoff.set(topic, backoff)
+    }
+    const expire = this._now() + interval
+    const existingExpire = backoff.get(id) || 0
+    if (existingExpire < expire) {
+      backoff.set(id, expire)
+    }
+  }
+
+  /**
+   * Clear expired backoff expiries
+   * @returns {void}
+   */
+  _clearBackoff (): void {
+    // we only clear once every 15 ticks to avoid iterating over the maps too much
+    if (this.heartbeatTicks % 15 !== 0) {
```

Comment:
Where does the 15 come from? Should this be a constant/configuration option?

Reply:
Also, should we reset it instead?

Reply:
15 is just pulled from go-libp2p-pubsub. I'll pull it into a constant.

We can't reset it, at least not here, because other features will be relying on the ticks in a similar way (eg: https://github.com/ChainSafe/js-libp2p-gossipsub/pull/96/files#diff-86c7c8f2b553297cee358fb131db4f7bR239)
In go-libp2p-pubsub, this counter is never reset, even on `stop`. Shouldn't be a problem for us though, even at a heartbeat every 100ms, we should have over 10 million years before getting out of safe integer range.